packages/gatsby/src/utils/cache.ts

```diff
@@ -0,0 +1,77 @@
+import manager, { Store, StoreConfig, CachingConfig } from "cache-manager"
+import fs from "fs-extra"
+import fsStore from "cache-manager-fs-hash"
+import path from "path"
+
+const MAX_CACHE_SIZE = 250
+const TTL = Number.MAX_SAFE_INTEGER
+
+interface ICacheProperties {
+  name?: string
+  store?: Store
+}
+
+class Cache {
+  public name: string
+  public store: Store
+  public cache?: manager.Cache
+
+  constructor({ name = `db`, store = fsStore }: ICacheProperties = {}) {
+    this.name = name
+    this.store = store
+  }
+
+  get directory(): string {
+    return path.join(process.cwd(), `.cache/caches/${this.name}`)
+  }
+
+  init(): Cache {
+    fs.ensureDirSync(this.directory)
+
+    const configs: StoreConfig[] = [
+      {
+        store: `memory`,
+        max: MAX_CACHE_SIZE,
+        ttl: TTL,
+      },
+      {
+        store: this.store,
+        ttl: TTL,
+        options: {
+          path: this.directory,
+          ttl: TTL,
+        },
+      },
+    ]
+
+    const caches = configs.map(cache => manager.caching(cache))
+
+    this.cache = manager.multiCaching(caches)
+
+    return this
+  }
+
+  get<T = unknown>(key): Promise<T | undefined> {
+    return new Promise(resolve => {
+      // eslint-disable-next-line no-unused-expressions
+      this.cache?.get<T>(key, (err, res) => {
+        resolve(err ? undefined : res)
+      })
+    })
+  }
+
+  set<T>(
+    key: string,
+    value: T,
+    args: CachingConfig = { ttl: TTL }
+  ): Promise<T | undefined> {
```

Comment:
We had a goal to avoid using generics as much as possible. Do you think it's possible to do here? I'm not totally familiar with the cache module yet. Is the value supposed to be any specific expected types? I would look through the codebase and see when we use this if we align on any specific types.

Reply:
Generics are unavoidable if you're typing Promises. I'm also using them for the cache getters and setters, because the return type depends on the input type.